### ISSUE TITLE
C++/CLR IJwhost.dll Exception on windows server 2012 (works on Win 10)

### ISSUE THREAD

**Issue by fkahhaleh** [dotnet/runtime#71812]
### Description

Hi,

I've been trying to trace an issue with an application I am writing. To control variables, I've created a POC solution that mimics our larger solution and was successful in replicating the error.

I have a Native C++ project ---Calls--> C++/CLR as Bridge --> Calls C# .Net DLL.
CLR is set to (/clr:netcore) and specifies .netcore3.1. The C# portion of course is compiled for .Net Core 3.1 as well.

On my dev machine, running windows 10, the console (native c++) application runs correctly without issue.
However once I place it on our server (Win Server 2012) I get an error.
Event Viewer states:
```
Faulting application name: POC_MOCK.exe, version: 0.0.0.0, time stamp: 0x62c75fef
Faulting module name: ucrtbase.DLL, version: 10.0.14393.2990, time stamp: 0x5caeb859
Exception code: 0xc0000409
Fault offset: 0x000884cb
Faulting process id: 0x1c9c
Faulting application start time: 0x01d8927e8d0fe66f
Faulting application path: D:\test\POC_MOCK.exe
Faulting module path: C:\Windows\SYSTEM32\ucrtbase.DLL
Report Id: cad1e77c-fe71-11ec-80fe-000c297ddf34
Faulting package full name: 
Faulting package-relative application ID: 
```

I took the crash dump and analyzed it via winDbg preview and got the following. (for both the POC and Actual production code that fails)
```
STACK_TEXT:  
00e2fc60 5e00c672     e06d7363 00000001 00000003 KERNELBASE!RaiseException+0x48
00e2fc90 5e001405     00e2fcb0 5e015750 00f58fe8 Ijwhost!_CxxThrowException+0x66
00e2fcdc 5e009d4c     013c0667 00000000 00f5dc20 Ijwhost!start_runtime_and_get_target_address+0x138
00e2fcf0 5e0b1fc5     c42602c1 00f5a430 00f5dc20 Ijwhost!start_runtime_thunk_stub+0xc
00e2fd10 01201041     9b12698d 73d9cf33 9b12698d CPPWrapper!CPPWrapper::CPPWrapper+0x45
00e2fd30 0120159c     00000001 00f5a430 00f58fe8 POC_MOCK!main+0x41
00e2fd78 74bd6a14     7f218000 74bd69f0 eefe366c POC_MOCK!__scrt_common_main_seh+0xfa
00e2fd8c 7712a9ff     7f218000 ed2f7874 00000000 kernel32!BaseThreadInitThunk+0x24
00e2fdd4 7712a9ca     ffffffff 7710fdc7 00000000 ntdll!__RtlUserThreadStart+0x2f
00e2fde4 00000000     01201624 7f218000 00000000 ntdll!_RtlUserThreadStart+0x1b

FAULTING_SOURCE_LINE:  D:\a\_work\1\s\src\installer\corehost\cli\ijwhost\ijwthunk.cpp
FAULTING_SOURCE_FILE:  D:\a\_work\1\s\src\installer\corehost\cli\ijwhost\ijwthunk.cpp
FAULTING_SOURCE_LINE_NUMBER:  153

FAULTING_SOURCE_CODE:  
No source found for 'D:\a\_work\1\s\src\installer\corehost\cli\ijwhost\ijwthunk.cpp'

SYMBOL_NAME:  ijwhost!start_runtime_and_get_target_address+138
MODULE_NAME: Ijwhost
IMAGE_NAME:  Ijwhost.dll
STACK_COMMAND:  .cxr 0xe2f920 ; kb
FAILURE_BUCKET_ID:  FAIL_FAST_FATAL_APP_EXIT_CPP_EXCEPTION_c0000409_Ijwhost.dll!start_runtime_and_get_target_address
OS_VERSION:  6.3.9600.18217
BUILDLAB_STR:  winblue_ltsb
OSPLATFORM_TYPE:  x86
OSNAME:  Windows 8.1
IMAGE_VERSION:  5.0.1722.21314
FAILURE_ID_HASH:  {b2047410-307d-cb83-9a4d-78b763edf2d5}
Followup:     MachineOwner
```

is Ijwhost unable to get the right proc address after loading the managed dll?
I've tried to find some infomration online, but most resources mentioned are regarding not finding ijwhost.dll to load.

### Reproduction Steps

Native C++ application (N) that references a C++ CLR application that is acting as a bridge to reference a C# ClassLibrary (DLL).

### Expected behavior

Application should run on operating systems windows 8.1 and newer

### Actual behavior

Ijwhost.dll exception occurs as shown in the crash dump.

### Configuration

Native C++ compiled with vc143 and Win10 SDK 10.022.
CLR targeting .Net Core 3.1
C# Class library targeting .Net Core 3.1
.Net Core 3.1 confirmed to be installed on the server.
supporting .runtime.json files are there.
Everything is compiled for x86.

**Comment by AaronRobinsonMSFT** [comment]
@fkahhaleh Could you share your reproduction project with us?

**Comment by vitek-karas** [comment]
Can you try running a normal .NET app on the target server? There are some additional requirements for the .NET runtime to work on Windows 2012: https://docs.microsoft.com/en-us/dotnet/core/install/windows?tabs=net60#additional-deps

To get more information you could also set COREHOST_TRACE=1, COREHOST_TRACEFILE=host.txt and repro the problem. The `host.txt` should contain more information about the failure.

**Comment by AaronRobinsonMSFT** [comment]
> it's a bit intertangled as you'd guess and my P.O.C project might not be enough

The POC is often enough to get me down the right path, we can start with that.

**Comment by AaronRobinsonMSFT** [comment]
> I switched over to .Net Framework 4.8 out of curiosity and it would work fine, my CLR DLL is loaded and is working as expected as a proxy/wrapper for the C# DLL.

Comparing .NET Framework C++/CLI and .NET Core 3.1 C++/CLI shouldn't be done. The hosting of between the two is fundamentally different.

**Comment by AaronRobinsonMSFT** [comment]
@fkahhaleh I am unable to reproduce this on Windows 2012 with the following design. This is targeting .NET Core 3.1 for x86.

## Native EXE
```cpp
#include <iostream>
#include "Proxy.h" // Header for the C++/CLI assembly

int main()
{
    auto p = new Proxy();
    p->SayHello();

    return EXIT_SUCCESS;
}
```

## C++/CLI assembly

#### Proxy.h
```hpp
#pragma once

#if AS_LIB
#define DllExport   __declspec( dllexport )
#else
#define DllExport   __declspec( dllimport )
#endif

class DllExport Proxy
{
public:
    void SayHello();
};
```

#### Proxy.cpp
```cpp
#include "Proxy.h"

void Proxy::SayHello()
{
    net_lib::Class^ c = gcnew net_lib::Class();
    c->Hello();
}
```

## C# assembly
```csharp
using System;

namespace net_lib
{
    public class Class
    {
        public void Hello()
        {
            Console.WriteLine("Hello from C#!");
        }
    }
}
```

**Comment by fkahhaleh** [comment]
@AaronRobinsonMSFT 
Sorry for the delay in any feedback on my end.
I've taken the above code and to my surprise it is working without any issue.

I've hit an issue with the POC project that I have on my end, not sure what's going on with it but getting a
_- System.BadImageFormatException: An attempt was made to load a program with an incorrect format. (Exception from HRESULT: 0x8007000B)_
As if I am mixing 32 and 64bit (or AnyCPU), although I am not. So I'll take the remainder of the day to work that out before
I share feedback and hopefully provide the project for you to test if possible.

Thank you!

**Comment by fkahhaleh** [comment]
Out of curiosity, does this look ok? 
It was missing the Core Runtime -3.1.26 (x86) and your code above worked in both cases before and after installing it.
![image](https://user-images.githubusercontent.com/9301486/180485240-716eca36-f49b-4893-99f6-2f334e4ed2e4.png)

**Comment by AaronRobinsonMSFT** [comment]
> It was missing the Core Runtime -3.1.26 (x86) and your code above worked in both cases before and after installing it.

This would appear to mean there is indeed a confusion during run-time.  Were you able to collect the logs that @vitek-karas mention in https://github.com/dotnet/runtime/issues/71812#issuecomment-1188034439

**Comment by fkahhaleh** [comment]
@AaronRobinsonMSFT 
Thank you for your follow up and support guys.

I am not sure what to say about the issue here, because when I tested the skeleton code you provided and it worked, I was motivated to start fresh and work my way up to the code I wanted/had.
It was very difficult for me to have a clear history of changes I made given I was implementing a clr host approach to unblock
the team while we figure out how to get this to work. (i.e hosting clr, loading the managed c# DLL and obtaining function pointers ..etc which surprisingly got to work fairly quickly)

So last week I went back to your code above, and found that I might have had an issue of how I was structuring the code. I believe it was a combination between how I was initializing the managed C++ code and a misplaced  #pragma managed. 
It beats  me how it was working on my development machine and not on the server, so perhaps doing many changes and the server environment not having the x86 version installed were at play here!

After starting over with a clean slate, I was able to get everything working.

I'll close the issue . Hoping that the basic code above would be helpful to anyone trying to implement a native to managed C++ bridge as a quick confirmation method that it SHOULD work, otherwise something else besides that code is at fault. :)

Thank you again.
Fadi
